File: get-variable-data/get-variable-data.R
########################################################################################
# GMWO API Example - Getting variable data                                             #
# For more explanation of the steps, see readme.md                                     #
#                                                                                      #  
# Prerequisites:                                                                       #  
# - R installed                                                                        #
# - RStudio installed                                                                  #    
# - a GMWO access token, see https://model.oxfordeconomics.com/api/docs#authentication #
########################################################################################


# import modules
require('httr')
require('jsonlite')
require('comprehenr')

# program constants
# note: ACCESS_TOKEN must be set below
#
ACCESS_TOKEN <- ''
AUTHORIZATION_HEADER_VALUE <- paste('Bearer', ACCESS_TOKEN,sep = " ")
BASE <- "https://model.oxfordeconomics.com/api"


get_Request<- function(endpoint, authorization_header_value)
{
    response <-GET(endpoint, add_headers("Authorization" = authorization_header_value))

    json_response <- parse_json(response)

    return(json_response)
}

get_Resources <- function(resourceId)
{
    endpoint <- paste(BASE,'/v1/resources/', resourceId, sep="" )    
    
    return(get_Request(endpoint, AUTHORIZATION_HEADER_VALUE))
}

get_Indicators <- function(forecastId)
{
    endpoint <- paste(BASE,'/v1/forecast-contents/', forecastId, '/indicators',sep="")

    return(get_Request(endpoint, AUTHORIZATION_HEADER_VALUE))
}

get_Locations <- function(forecastId)
{
    endpoint <- paste(BASE, '/v1/forecast-contents/', forecastId, '/groups', sep="")
   
    return(get_Request(endpoint, AUTHORIZATION_HEADER_VALUE))
}

get_VariableData <- function(forecastId, locations, indicators)
{
    body <- to_list(for(group in locations) for(indicator in indicators) list(IndicatorCode=indicator$Code, GroupCode=group$Code))

    endpoint <- paste(BASE, '/v1/forecast-contents/', forecastId, '/model-variables', sep="")
    response <- POST(
        endpoint,
        body = body, 
        add_headers("Authorization" = AUTHORIZATION_HEADER_VALUE),
        query= list(onlyProps = "Values,Metadata"),
        encode="json"
    )
    data <- parse_json(response)
    
    return(data)
}

# get a list of forecasts
resources <- get_Resources("oxford-economics/releases/GEM")

# get the first forecast from the macro releases
forecastId <- resources$Children[[1]]$Id[1]

# get list of indicators
indicators <- head(get_Indicators(forecastId), 5) # use only first 5 indicators

# get list of locations(groups)
locations <- head(get_Locations(forecastId), 5) # use only first 5 locations (groups)

# get variable data
raw_data <- get_VariableData(forecastId, locations, indicators)

print(raw_data)

File: get-variable-data/get-variable-data.py
########################################################################################
# GMWO API Example - Getting variable data                                             #
# For more explanation of the steps, see readme.md                                     #
#                                                                                      #  
# Prerequisites:                                                                       #  
# - Python 3 installed                                                                 #
# - 'requests' package installed                                                       #    
# - a GMWO access token, see https://model.oxfordeconomics.com/api/docs#authentication #
########################################################################################

import requests

base_url = 'https://model.oxfordeconomics.com/api'
access_token = <insert your access token here>
headers = {'Authorization' : f'Bearer {access_token}'}

# Discover the id of a Global Economic Model release forecast
gemFolderResponse = requests.get(f"{base_url}/v1/resources/oxford-economics/releases/GEM", headers=headers)
gemFolder = gemFolderResponse.json()
forecastId = gemFolder["Children"][0]["Id"] # Pick first forecast

# Get all indicators for a forecast
indicatorsResponse = requests.get(f"{base_url}/v1/forecast-contents/{forecastId}/indicators", headers=headers)
indicators = indicatorsResponse.json()

 # Get all groups (locations) for a forecast
groupsResponse = requests.get(f"{base_url}/v1/forecast-contents/{forecastId}/groups", headers=headers)
groups = groupsResponse.json()

# Get variable values
indicatorCodes = map(lambda indicator: indicator["Code"], indicators[0:4]) # Pick first 4 indicators
groupCodes = map(lambda group: group["Code"], groups[0:4]) # Pick first 4 groups
variableSelection = [ { "IndicatorCode": indicatorCode, "GroupCode": groupCode } for indicatorCode in indicatorCodes for groupCode in groupCodes]
variablesResponse = requests.post(f"{base_url}/v1/forecast-contents/{forecastId}/variables?onlyProps=Values,Metadata", headers=headers, json=variableSelection)

print(variablesResponse.json())
